File: java/rest-simple/src/test/java/com/example/restsimple/StudentServiceImplTest.java
package com.example.restsimple;

import com.example.restsimple.model.Student;
import com.example.restsimple.repository.StudentRepository;
import com.example.restsimple.repository.UnitOfWork;
import com.example.restsimple.repository.UnitOfWork.TransactionCode;
import com.example.restsimple.repository.UpdateStudent;
import com.example.restsimple.service.StudentServiceImpl;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static junit.framework.TestCase.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class StudentServiceImplTest {

    @Mock
    private StudentRepository studentRepository;

    @Mock
    private UnitOfWork unitOfWork;

    @InjectMocks
    private StudentServiceImpl studentService;

    @Test
    public void testGetAllStudents() {
        List<Student> students = Arrays.asList(new Student());
        when(unitOfWork.executeInTransaction(any())).thenReturn(students);

        List<Student> result = studentService.getAllStudents();

        assertEquals(students, result);
    }

    @Test
    public void testCreateStudent() {
        Student student = new Student();
        when(unitOfWork.executeInTransaction(any())).thenReturn(student);

        Student result = studentService.createStudent(student);

        assertEquals(student, result);
    }

    @Test
    public void testUpdateStudent() {
        String id = "1";
        UpdateStudent updateStudent = new UpdateStudent("John", "Doe");
        Student existingStudent = new Student();
        when(unitOfWork.executeInTransaction(any(TransactionCode.class))).thenAnswer(invocation -> {
            TransactionCode<Student> code = invocation.getArgument(0);
            return code.apply();
        });
        when(studentRepository.findByIdColumn(id)).thenReturn(Optional.of(existingStudent));

        Optional<Student> result = studentService.updateStudent(id, updateStudent);

        assertEquals(existingStudent, result.get());
        assertEquals("John", result.get().getName());
        assertEquals("Doe", result.get().getLastName());
    }

    @Test
    public void testUpdateStudent_NotFound() {
        String id = "1";
        UpdateStudent updateStudent = new UpdateStudent("John", "Doe");

        when(unitOfWork.executeInTransaction(any(TransactionCode.class))).thenAnswer(invocation -> {
            TransactionCode<Optional<Student>> code = invocation.getArgument(0);
            return code.apply();
        });

        when(studentRepository.findByIdColumn(id)).thenReturn(Optional.empty());

        Optional<Student> result = studentService.updateStudent(id, updateStudent);

        assertFalse(result.isPresent());
    }


    @Test
    public void testDeleteStudent() {
        String id = "1";
        Student student = new Student();
        when(unitOfWork.executeInTransaction(any(TransactionCode.class))).thenAnswer(invocation -> {
            TransactionCode<Student> code = invocation.getArgument(0);
            return code.apply();
        });
        when(studentRepository.findByIdColumn(id)).thenReturn(Optional.of(student));

        Optional<Student> result = studentService.deleteStudent(id);

        assertEquals(student, result.get());
        verify(studentRepository).delete(student);
    }


    @Test
    public void testDeleteStudent_NotFound() {
        String id = "1";
        when(unitOfWork.executeInTransaction(any(TransactionCode.class))).thenAnswer(invocation -> {
            TransactionCode<Optional<Student>> code = invocation.getArgument(0);
            return code.apply();
        });
        when(studentRepository.findByIdColumn(id)).thenReturn(Optional.empty());

        Optional<Student> result = studentService.deleteStudent(id);

        assertFalse(result.isPresent());
    }

    @Test
    public void testFindStudentById() {
        String id = "1";
        Student student = new Student();
        when(unitOfWork.executeInTransaction(any(TransactionCode.class))).thenAnswer(invocation -> {
            TransactionCode<Optional<Student>> code = invocation.getArgument(0);
            return code.apply();
        });
        when(studentRepository.findByIdColumn(id)).thenReturn(Optional.of(student));

        Optional<Student> result = studentService.findStudentById(id);

        assertEquals(student, result.get());
    }
}

File: kotlin/rest-simple/src/test/kotlin/com/example/restsimple/StudentServiceImplTest.kt
package com.example.restsimple

import com.example.restsimple.model.Student
import com.example.restsimple.repository.StudentRepository
import com.example.restsimple.repository.UnitOfWork
import com.example.restsimple.repository.UpdateStudent
import com.example.restsimple.service.StudentServiceImpl
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.kotlin.anyOrNull
import org.springframework.boot.test.context.SpringBootTest
import java.time.LocalDateTime
import java.util.*
import org.mockito.Mockito.`when` as whenever

@SpringBootTest
class StudentServiceImplTest {

    @Mock
    private lateinit var studentRepository: StudentRepository

    @Mock
    private lateinit var unitOfWork: UnitOfWork

    private lateinit var studentService: StudentServiceImpl

    @BeforeEach
    fun setUp() {
        MockitoAnnotations.openMocks(this)
        studentService = StudentServiceImpl(studentRepository, unitOfWork)
    }

    @Test
    fun `test get all students`() {
        val students = mutableListOf<Student?>(Student("1", "John", "Doe", LocalDateTime.now()))

        // Mock executeInTransaction to directly execute the passed lambda.
        whenever(unitOfWork.executeInTransaction<MutableList<Student?>>(anyOrNull())).thenAnswer {
            (it.arguments[0] as? Function0<MutableList<Student?>>)?.invoke()
        }

        whenever(studentRepository.findAll()).thenReturn(students)

        val result = studentService.allStudents

        assert(result.size == 1)
        assert(result[0]?.name == "John")
    }

    @Test
    fun `test create student`() {
        val student = Student("1", "John", "Doe", LocalDateTime.now())

        whenever(unitOfWork.executeInTransaction<Student?>(anyOrNull())).thenAnswer {
            (it.arguments[0] as? Function0<Student?>)?.invoke()
        }

        whenever(studentRepository.save(student)).thenReturn(student)

        val result = studentService.createStudent(student)

        assert(result?.name == "John")
    }

    @Test
    fun `test update student`() {
        val id = "1"
        val updatedStudent = UpdateStudent("Jane", "Doe")
        val existingStudent = Student("1", "John", "Doe", LocalDateTime.now())

        whenever(studentRepository.findByIdColumn(id)).thenReturn(Optional.of(existingStudent))
        whenever(unitOfWork.executeInTransaction<Optional<out Student?>>(anyOrNull())).thenAnswer {
            (it.arguments[0] as? Function0<Optional<out Student?>>)?.invoke()
        }

        val result = studentService.updateStudent(id, updatedStudent)

        assert(result.isPresent)
        assert(result.get()?.name == "Jane")
    }

    @Test
    fun `test delete student`() {
        val id = "1"
        val student = Student("1", "John", "Doe", LocalDateTime.now())

        whenever(studentRepository.findByIdColumn(id)).thenReturn(Optional.of(student))
        whenever(unitOfWork.executeInTransaction<Optional<out Student?>>(anyOrNull())).thenAnswer {
            (it.arguments[0] as? Function0<Optional<out Student?>>)?.invoke()
        }

        val result = studentService.deleteStudent(id)

        assert(result.isPresent)
    }

    @Test
    fun `test find student by id`() {
        val id = "1"
        val student = Student("1", "John", "Doe", LocalDateTime.now())

        whenever(studentRepository.findByIdColumn(id)).thenReturn(Optional.of(student))
        whenever(unitOfWork.executeInTransaction<Optional<Student>>(anyOrNull())).thenAnswer {
            (it.arguments[0] as? Function0<Optional<Student>>)?.invoke()
        }

        val result = studentService.findStudentById(id)

        assert(result.isPresent)
        assert(result.get().name == "John")
    }
}

